lib/ClangImporter/SwiftDeclSynthesizer.cpp

```diff
@@ -2534,128 +2534,192 @@ SwiftDeclSynthesizer::makeDefaultArgument(const clang::ParmVarDecl *param,
 
 // MARK: C++ foreign reference type constructors
 
-clang::CXXMethodDecl *
+llvm::SmallVector<clang::CXXMethodDecl *, 4>
 SwiftDeclSynthesizer::synthesizeStaticFactoryForCXXForeignRef(
     const clang::CXXRecordDecl *cxxRecordDecl) {
 
   clang::ASTContext &clangCtx = cxxRecordDecl->getASTContext();
   clang::Sema &clangSema = ImporterImpl.getClangSema();
 
   clang::QualType cxxRecordTy = clangCtx.getRecordType(cxxRecordDecl);
+  clang::SourceLocation cxxRecordDeclLoc = cxxRecordDecl->getLocation();
 
-  clang::CXXConstructorDecl *defaultCtorDecl = nullptr;
-  for (clang::CXXConstructorDecl *ctor : cxxRecordDecl->ctors()) {
-    if (ctor->parameters().empty() && !ctor->isDeleted() &&
-        ctor->getAccess() != clang::AS_private &&
-        ctor->getAccess() != clang::AS_protected) {
-      defaultCtorDecl = ctor;
-      break;
+  llvm::SmallVector<clang::CXXConstructorDecl *, 4> ctorDeclsForSynth;
+  for (clang::CXXConstructorDecl *ctorDecl : cxxRecordDecl->ctors()) {
+    if (ctorDecl->isDeleted() || ctorDecl->getAccess() == clang::AS_private ||
+        ctorDecl->getAccess() == clang::AS_protected ||
+        ctorDecl->isCopyOrMoveConstructor() || ctorDecl->isVariadic())
+      continue;
+    bool hasDefaultArg = false;
```

Comment:
You don't need a loop to check if a function has a default argument. It is enough to only look at the last parameter (if there is at least one). This is because in C++ if a parameter has default argument, all the following ones also must have default arguments. So you can get rid of this loop.